drivers/clock_control/nrf_clock_calibration.c

```diff
@@ -0,0 +1,272 @@
+	}
+
+	cal_state = CAL_IDLE;
+
+	irq_unlock(key);
+
+	calib_skip_cnt = 0;
+
+	return ret;
+}
+
+bool nrf_clock_control_calibration_stop(struct device *dev)
+{
+	int key;
+	bool ret = true;
+
+	key = irq_lock();
+
+	nrf_clock_task_trigger(NRF_CLOCK_TASK_CTSTOP);
+	nrf_clock_event_clear(NRF_CLOCK_EVENT_CTTO);
+
+	/* If calibration is active then pend until completed.
+	 * Currently (and most likely in the future), LFCLK is never stopped so
+	 * it is not an issue.
+	 */
+	if (cal_state == CAL_ACTIVE) {
+		cal_state = CAL_ACTIVE_OFF;
+		ret = false;
+		LOG_INF("Stop when calibration is active");
+	} else {
+		cal_state = CAL_OFF;
+	}
+
+	irq_unlock(key);
+
+	return ret;
+}
+
+void nrf_clock_control_calibration_init(struct device *dev)
+{
+	/* Anomaly 36: After watchdog timeout reset, CPU lockup reset, soft
+	 * reset, or pin reset EVENTS_DONE and EVENTS_CTTO are not reset.
+	 */
+	nrf_clock_event_clear(NRF_CLOCK_EVENT_DONE);
+	nrf_clock_event_clear(NRF_CLOCK_EVENT_CTTO);
+
+	nrf_clock_int_enable(NRF_CLOCK_INT_DONE_MASK |
+			     NRF_CLOCK_INT_CTTO_MASK |
+			     NRF_CLOCK_INT_CTSTOPPED_MASK);
+	nrf_clock_cal_timer_timeout_set(
+			CONFIG_CLOCK_CONTROL_NRF_CALIBRATION_PERIOD);
+
+	if (CONFIG_CLOCK_CONTROL_NRF_CALIBRATION_MAX_SKIP != 0) {
+		temp_sensor = device_get_binding(TEMP_SENSOR_NAME);
+	}
+
+	hfclk_dev = dev;
+}
+
+/* Start calibration assuming that HFCLK XTAL is on. */
+static void start_calibration(void)
+{
+	cal_state = CAL_ACTIVE;
+	nrf_clock_task_trigger(NRF_CLOCK_TASK_CAL);
+	LOG_DBG("Start calibration.");
+	calib_skip_cnt = CONFIG_CLOCK_CONTROL_NRF_CALIBRATION_MAX_SKIP;
+}
+
+/* Restart calibration timer, release HFCLK XTAL. */
+static void to_idle(void)
+{
+	cal_state = CAL_IDLE;
+	clock_control_off(hfclk_dev, 0);
+	nrf_clock_task_trigger(NRF_CLOCK_TASK_CTSTART);
+}
+
+/* Convert sensor value to 0.25'C units. */
+static inline s16_t sensor_value_to_temp_unit(struct sensor_value *val)
+{
+	return (s16_t)(4 * val->val1 + val->val2 / 250000);
+}
+
+/* Function reads from temperature sensor and converts to float. */
+static s16_t get_temperature(void)
+{
+	struct sensor_value sensor_val;
+
+	sensor_sample_fetch(temp_sensor);
+	sensor_channel_get(temp_sensor, SENSOR_CHAN_DIE_TEMP, &sensor_val);
+
+	return sensor_value_to_temp_unit(&sensor_val);
+}
+
+/* Function determines if calibration should be performed based on temperature
+ * measurement. Function is called from system work queue context. It is
+ * reading temperature from TEMP sensor and compares with last measurement.
+ */
+static void measure_temperature(struct k_work *work)
+{
+	s16_t temperature;
+	s16_t diff;
+	int key;
+
+	temperature = get_temperature();
+	diff = temperature - prev_temperature;
+	diff = diff < 0 ? -diff : diff;
+	prev_temperature = temperature;
+
+	key = irq_lock();
+
+	if (cal_state != CAL_OFF) {
+		if (diff >= CONFIG_CLOCK_CONTROL_NRF_CALIBRATION_TEMP_DIFF) {
+			start_calibration();
+		} else {
+			LOG_DBG("Calibration skipped - stable temperature.");
```

Comment:
IMHO it might be a good idea to print the temperature difference here.

Reply:
done